### DISCUSSION TITLE
[DNS Client] utilization of otDnsClientResolveIp4Address method

### DISCUSSION THREAD

**Discussion by estelleOl** [openthread/openthread#8296] [Q&A] [answered]
Hello,

I try to use the otDnsClientResolveIp4Address method but it return the error OT_ERROR_NOT_FOUND (23).
When I debug in the code method in dns_client.cpp in the Client::StartQuery method the existence of a nat64Prefix on the leader board is checked and in my case it fails.
I have 2 boards, one is in leader mode and the other is in router mode. 

I have configured the prefix on my leader board (with a length of 64) like that : 
 /* Add a first Off Mesh Prefix */
 otIp6AddressFromString("2001:dead:beef:cafe::/64", &config[0].mPrefix.mPrefix);
 config[0].mPrefix.mLength = 64;

  status = otBorderRouterAddRoute(NULL, &config[0]);
        
   /* Add a second Off-Mesh Prefix */
   otIp6AddressFromString("2002:dead:beef:cafe::/64", &config[1].mPrefix.mPrefix);
   config[1].mPrefix.mLength = 64;

   status = otBorderRouterAddRoute(NULL, &config[1]);
        
I call otDnsClientResolveIp4Address method in the router board after having allowed nat64 in dns config setup like that : 
/* Set the default query config on DNS client */
defaultConfigSet.mResponseTimeout = 2000;
defaultConfigSet.mNat64Mode = OT_DNS_NAT64_ALLOW;
otDnsClientSetDefaultConfig(NULL, &defaultConfigSet);
	   
/* Send a DNS service instance resolution query for a given service instance */
status = otDnsClientResolveIp4Address (NULL, "Board2 Name.default.service.arpa.", dnsAddressCallback, NULL, NULL);

I think the prefix is not recognized as a nat64Prefix. I look in the code it seems that a function named GenerateNat64Prefix() in routing_manager.cpp can be used to declare a nat64Prefix. Is it the good way to do it ? This function is protected by the OPENTHREAD_CONFIG_BORDER_ROUTING_NAT64_ENABLE flag is that flag mandatory to use correctly the otDnsClientResolveIp4Address method ?

**Comment by jwhui** [comment, marked as answer]
A NAT64 prefix must be explicitly configured.

There are two border router build features to be aware of:
- `OT_NAT64_BORDER_ROUTING` enables the automatic configuration of a NAT64 prefix.
- `OT_NAT64_TRANSLATOR` enables the translation between IPv4 and IPv6.

**Reply by estelleOl** [reply]
If I do not have enabled those to builds feature, is it normal that otDnsClientResolveIp4Address return an error ?

**Reply by jwhui** [reply]
Yes, returning error is expected when a NAT64 prefix is not configured by a border router.

https://github.com/openthread/openthread/blob/d5333ca831a9942127a2c8d5370fe24dd278f7b1/src/core/net/dns_client.cpp#L723-L724

**Reply by estelleOl** [reply]
Ok , thank you for your answers.
